plugins/USBPrinting/USBPrinterOutputDeviceManager.py

```diff
@@ -170,48 +172,53 @@ def _getDefaultFirmwareName(self):
             Logger.log("e", "Could not find any firmware for machine %s.", machine_type)
             raise FileNotFoundError()
 
+    ##  Helper to identify serial ports (and scan for them)
     def _addRemovePorts(self, serial_ports):
         # First, find and add all new or changed keys
         for serial_port in list(serial_ports):
             if serial_port not in self._serial_port_list:
-                self.addConnectionSignal.emit(serial_port) #Hack to ensure its created in main thread
+                self.addUSBOutputDeviceSignal.emit(serial_port)  # Hack to ensure its created in main thread
                 continue
         self._serial_port_list = list(serial_ports)
 
-        connections_to_remove = []
-        for port, connection in self._printer_connections.items():
+        devices_to_remove = []
+        for port, device in self._usb_output_devices.items():
             if port not in self._serial_port_list:
-                connection.close()
-                connections_to_remove.append(port)
-
-        for port in connections_to_remove:
-            del self._printer_connections[port]
+                device.close()
+                devices_to_remove.append(port)
 
+        for port in devices_to_remove:
+            del self._usb_output_devices[port]
 
     ##  Because the model needs to be created in the same thread as the QMLEngine, we use a signal.
-    def addConnection(self, serial_port):
-        connection = PrinterConnection.PrinterConnection(serial_port)
-        connection.connect()
-        connection.connectionStateChanged.connect(self._onPrinterConnectionStateChanged)
-        connection.progressChanged.connect(self.progressChanged)
-        self._printer_connections[serial_port] = connection
-
-    def _onPrinterConnectionStateChanged(self, serial_port):
-        if self._printer_connections[serial_port].isConnected():
-            self.getOutputDeviceManager().addOutputDevice(self._printer_connections[serial_port])
-        else:
-            self.getOutputDeviceManager().removeOutputDevice(serial_port)
-        self.printerConnectionStateChanged.emit()
+    def addOutputDevice(self, serial_port):
+        device = USBPrinterOutputDevice.USBPrinterOutputDevice(serial_port)
+        device.connect()
+        device.connectionStateChanged.connect(self._onConnectionStateChanged)
+        device.progressChanged.connect(self.progressChanged)
```

Comment:
Ideally, you should connect to signals before calling a method that may cause those signals to be emitted, unless you intentionally want to ignore those signals. In this case, if device.connect() is fast enough (or uses blocking calls) you end up with not being informed about connection state changes.